Windows weirdness with GitHub runners and temporary directories

### Description

Beginning with commit 739faf364, `cargo uitest` fails on Windows when Clippy is cloned into a temporary directory in a GitHub workflow.

Specifically, the [5041_allow_dev_build](https://github.com/rust-lang/rust-clippy/tree/master/tests/ui-cargo/multiple_crate_versions/5041_allow_dev_build) test fails.

Here is a test repo that demonstrates the problem: https://github.com/smoelius/ui_test_test/actions/runs/6161347090/job/16720379793#step:2:1386

I think it has something to do with there being multiple paths to the temporary directory. In the logs, it is referred to as both of the following:
- `C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp...`
- `C:\\Users\\runneradmin\\AppData\\Local\\Temp\\tmp...`

Any idea what could be wrong?

cc: @oli-obk

### Version

_No response_

### Additional Labels

_No response_

Ran into this or something similar when upgrading to `ui_test` 0.21.2, it bisected to https://github.com/oli-obk/ui_test/commit/56513650253ca371b28d7f3e9d48e33a65c2c84a

We were doing that locally already so now the paths were pretty long https://github.com/rust-lang/rust-clippy/blob/33f49f33d6f8cec9a5772bc7f3a4e5f1de207b30/tests/compile-test.rs#L301-L307

```
FAILED TEST: tests\ui-cargo\multiple_crate_versions\5041_allow_dev_build\Cargo.toml
command: "D:\\Code\\Rust\\clippy\\target\\debug\\cargo-clippy.exe" "clippy" "--color" "never" "--quiet" "--target-dir" "\\\\?\\D:\\Code\\Rust\\clippy\\target\\ui_test_cargo\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build" "--manifest-path" "tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\Cargo.toml"
```

A manual `cargo check` gives the same error:

```
$ cargo check --manifest-path tests/ui-cargo/multiple_crate_versions/5041_allow_dev_build/Cargo.toml --target-dir '\\?\D:\Code\Rust\clippy\target\ui_test_cargo\tests\ui-cargo\multiple_crate_versions\5041_allow_dev_build\tests\ui-cargo\multiple_crate_versions\5041_allow_dev_build'
   Compiling kernel32-sys v0.2.2
error: linking with `link.exe` failed: exit code: 1181
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\Users\\Alex\\AppData\\Local\\Temp\\rustclX4HJk\\symbols.o" "\\\\?\\D:\\Code\\Rust\\clippy\\target\\ui_test_cargo\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\debug\\build\\kernel32-sys-c610472e08096568\\build_script_build-c610472e08096568.build_script_build.a21a107244a5edd0-cgu.0.rcgu.o" "\\\\?\\D:\\Code\\Rust\\clippy\\target\\ui_test_cargo\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\debug\\build\\kernel32-sys-c610472e08096568\\build_script_build-c610472e08096568.2npodab9y7a9vn5q.rcgu.o" "/LIBPATH:D:\\Code\\Rust\\clippy\\target\\ui_test_cargo\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\debug\\deps" "/LIBPATH:C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "D:\\Code\\Rust\\clippy\\target\\ui_test_cargo\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\debug\\deps\\libbuild-812cb315295294cd.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-5a4aac4b2cd4ff41.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-e340a5e8908c7529.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-fdf9fb6cf4e81109.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd_detect-e69fa3eee02c54a0.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-fab259a51b0964e5.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-0f32565f081356f1.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-51e10326d334b70c.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-6ff672e2205e01ba.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-977079263e47efdc.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-fa89d85bc9ed894d.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-f117a9417935b51d.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-b86a9ba077d96bbb.rlib" "C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-6856e76ba99f7e60.rlib" "kernel32.lib" "advapi32.lib" "bcrypt.lib" "kernel32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "kernel32.lib" "ws2_32.lib" "kernel32.lib" "msvcrt.lib" "/NXCOMPAT" "/LIBPATH:C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:\\\\?\\D:\\Code\\Rust\\clippy\\target\\ui_test_cargo\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\tests\\ui-cargo\\multiple_crate_versions\\5041_allow_dev_build\\debug\\build\\kernel32-sys-c610472e08096568\\build_script_build-c610472e08096568.exe" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\Alex\\.rustup\\toolchains\\nightly-2023-10-06-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis"
  = note: LINK : fatal error LNK1181: cannot open input file '.obj'


error: could not compile `kernel32-sys` (build script) due to previous error
```

Maybe it was already close to whatever limit (path length, command line length?) it's running into and the temporary directory pushed it over the edge. #11645 should paper over this by removing all the canonicalization

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -29,7 +29,7 @@ color-print = "0.3.4" # Sync version with Cargo
 anstream = "0.5.0"
 
 [dev-dependencies]
-ui_test = "0.20"
+ui_test = "0.21.2"
 tester = "0.9"
 regex = "1.5"
 toml = "0.7.3"

--- a/tests/compile-test.rs
+++ b/tests/compile-test.rs
@@ -105,27 +105,20 @@ static EXTERN_FLAGS: LazyLock<Vec<String>> = LazyLock::new(|| {
 // whether to run internal tests or not
 const RUN_INTERNAL_TESTS: bool = cfg!(feature = "internal");
 
-fn canonicalize(path: impl AsRef<Path>) -> PathBuf {
-    let path = path.as_ref();
-    fs::create_dir_all(path).unwrap();
-    fs::canonicalize(path).unwrap_or_else(|err| panic!("{} cannot be canonicalized: {err}", path.display()))
-}
-
 fn base_config(test_dir: &str) -> (Config, Args) {
     let mut args = Args::test().unwrap();
     args.bless |= var_os("RUSTC_BLESS").is_some_and(|v| v != "0");
 
+    let target_dir = PathBuf::from(var_os("CARGO_TARGET_DIR").unwrap_or_else(|| "target".into()));
     let mut config = Config {
         mode: Mode::Yolo {
             rustfix: ui_test::RustfixMode::Everything,
         },
-        stderr_filters: vec![(Match::PathBackslash, b"/")],
-        stdout_filters: vec![],
         filter_files: env::var("TESTNAME")
             .map(|filters| filters.split(',').map(str::to_string).collect())
             .unwrap_or_default(),
         target: None,
-        out_dir: canonicalize(var_os("CARGO_TARGET_DIR").unwrap_or_else(|| "target".into())).join("ui_test"),
+        out_dir: target_dir.join("ui_test"),
         ..Config::rustc(Path::new("tests").join(test_dir))
     };
     config.with_args(&args, /* bless by default */ false);
@@ -168,19 +161,13 @@ fn run_ui() {
     config
         .program
         .envs
-        .push(("CLIPPY_CONF_DIR".into(), Some(canonicalize("tests").into())));
-
-    let quiet = args.quiet;
+        .push(("CLIPPY_CONF_DIR".into(), Some("tests".into())));
 
     ui_test::run_tests_generic(
         vec![config],
         ui_test::default_file_filter,
         ui_test::default_per_file_config,
-        if quiet {
-            status_emitter::Text::quiet()
-        } else {
-            status_emitter::Text::verbose()
-        },
+        status_emitter::Text::from(args.format),
     )
     .unwrap();
 }
@@ -194,17 +181,12 @@ fn run_internal_tests() {
     if let OutputConflictHandling::Error(err) = &mut config.output_conflict_handling {
         *err = "cargo uitest --features internal -- -- --bless".into();
     }
-    let quiet = args.quiet;
 
     ui_test::run_tests_generic(
         vec![config],
         ui_test::default_file_filter,
         ui_test::default_per_file_config,
-        if quiet {
-            status_emitter::Text::quiet()
-        } else {
-            status_emitter::Text::verbose()
-        },
+        status_emitter::Text::from(args.format),
     )
     .unwrap();
 }
@@ -212,22 +194,9 @@ fn run_internal_tests() {
 fn run_ui_toml() {
     let (mut config, args) = base_config("ui-toml");
 
-    config.stderr_filters = vec![
-        (
-            Match::Exact(
-                canonicalize("tests")
-                    .parent()
-                    .unwrap()
-                    .to_string_lossy()
-                    .as_bytes()
-                    .to_vec(),
-            ),
-            b"$DIR",
-        ),
-        (Match::Exact(b"\\".to_vec()), b"/"),
-    ];
-
-    let quiet = args.quiet;
+    config
+        .stderr_filters
+        .push((Match::from(env::current_dir().unwrap().as_path()), b"$DIR"));
 
     ui_test::run_tests_generic(
         vec![config],
@@ -238,11 +207,7 @@ fn run_ui_toml() {
                 .envs
                 .push(("CLIPPY_CONF_DIR".into(), Some(path.parent().unwrap().into())));
         },
-        if quiet {
-            status_emitter::Text::quiet()
-        } else {
-            status_emitter::Text::verbose()
-        },
+        status_emitter::Text::from(args.format),
     )
     .unwrap();
 }
@@ -270,22 +235,9 @@ fn run_ui_cargo() {
     });
     config.edition = None;
 
-    config.stderr_filters = vec![
-        (
-            Match::Exact(
-                canonicalize("tests")
-                    .parent()
-                    .unwrap()
-                    .to_string_lossy()
-                    .as_bytes()
-                    .to_vec(),
-            ),
-            b"$DIR",
-        ),
-        (Match::Exact(b"\\".to_vec()), b"/"),
-    ];
-
-    let quiet = args.quiet;
+    config
+        .stderr_filters
+        .push((Match::from(env::current_dir().unwrap().as_path()), b"$DIR"));
 
     let ignored_32bit = |path: &Path| {
         // FIXME: for some reason the modules are linted in a different order for this test
@@ -297,20 +249,8 @@ fn run_ui_cargo() {
         |path, config| {
             path.ends_with("Cargo.toml") && ui_test::default_any_file_filter(path, config) && !ignored_32bit(path)
         },
-        |config, path, _file_contents| {
-            config.out_dir = canonicalize(
-                std::env::current_dir()
-                    .unwrap()
-                    .join("target")
-                    .join("ui_test_cargo/")
-                    .join(path.parent().unwrap()),
-            );
-        },
-        if quiet {
-            status_emitter::Text::quiet()
-        } else {
-            status_emitter::Text::verbose()
-        },
+        |_config, _path, _file_contents| {},
+        status_emitter::Text::from(args.format),
     )
     .unwrap();
 }

